mjpc/planners/cem/planner.cc

```diff
@@ -0,0 +1,554 @@
+  const std::shared_lock<std::shared_mutex> lock(mtx_);
+  if (use_previous) {
+    previous_policy.Action(action, state, time);
+  } else {
+    policy.Action(action, state, time);
+  }
+}
+
+// update policy via resampling
+void CEMPlanner::UpdateNominalPolicy(int horizon) {
+  // dimensions
+  int num_spline_points = candidate_policy[winner].num_spline_points;
+
+  // set time
+  double nominal_time = time;
+  double time_shift = mju_max(
+      (horizon - 1) * model->opt.timestep / (num_spline_points - 1), 1.0e-5);
+
+  // temp variables to help with averaging
+  std::vector<double> temp_avg(model->nu);  // holds avg control action over elites for each spline point t
+  std::vector<std::vector<double>> temp_elite_actions;
+  temp_elite_actions.resize(n_elites, std::vector<double>(model->nu, 0.0)); // shape=(n_elites, nu)
+
+  variance.assign(model->nu * num_spline_points, 0.0);  // reset variance to 0
+
+  // get spline points
+  for (int t = 0; t < num_spline_points; t++) {
+    times_scratch[t] = nominal_time;
+    temp_avg.assign(model->nu, 0.0);  // reset temp_avg to 0
+
+    // get the actions of the top n_elites policies and average them
+    // also update a variance parameter too
+    for (int i = 0; i < n_elites; i++) {
+      candidate_policy[trajectory_order[i]].Action(  // trajectory_order[i] is the (i+1)^th best trajectory
+        DataAt(temp_elite_actions[i], 0),  // copies the action from the cand policy into the i^th control vector
+        nullptr,
+        nominal_time
+      );
+      for (int k = 0; k < model->nu; k++) {
+        temp_avg[k] += temp_elite_actions[i][k] / n_elites;
+      }
+    }
+
+    // computing the sample variance of the control parameters
+    for (int k = 0; k < model->nu; k++) {
+      for (int i = 0; i < n_elites; i++) {
+        // [DEBUG] which way are params indexed?
+        variance[t * model->nu + k] += std::pow(temp_elite_actions[i][k] - temp_avg[k], 2) / (n_elites - 1);
+        // variance[k * num_spline_points + t] += std::pow(temp_elite_actions[i][k] - temp_avg[k], 2) / (n_elites - 1);
+      }
+    }
+
+    // [DEBUG] print the standard deviations
+    // ********************************************************************************************************
+    // for (int k = 0; k < model->nu; k++) {
+    //   std::cout << "stdev[" << t << "][" << k << "] = " << std::sqrt(variance[t * model->nu + k]) << std::endl;
+    // }
+    // ********************************************************************************************************
+
+    // assigning the averaged parameters to parameters_scratch
+    std::copy(temp_avg.begin(), temp_avg.end(), parameters_scratch.begin() + t * model->nu);
+    nominal_time += time_shift;
+  }
+
+  // if first iter, set variance to the initial variance
+  if (first_iter) {
+    variance.assign(model->nu * num_spline_points, std::pow(noise_exploration, 2));
+    first_iter = false;
+  }
+  first_iter = false;
```

Comment:
shouldn't need this outside the if statement

Reply:
thanks, that was a stray